File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/sparse/BsrToDenseBp.java
/*
 *  ******************************************************************************
 *  *
 *  *
 *  * This program and the accompanying materials are made available under the
 *  * terms of the Apache License, Version 2.0 which is available at
 *  * https://www.apache.org/licenses/LICENSE-2.0.
 *  *
 *  * See the NOTICE file distributed with this work for additional
 *  * information regarding copyright ownership.
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 *  * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 *  * License for the specific language governing permissions and limitations
 *  * under the License.
 *  *
 *  * SPDX-License-Identifier: Apache-2.0
 *  *****************************************************************************
 */

package org.nd4j.linalg.api.ops.impl.sparse;

import org.nd4j.autodiff.samediff.SDVariable;
import org.nd4j.autodiff.samediff.SameDiff;
import org.nd4j.linalg.api.buffer.DataType;
import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.api.ops.DynamicCustomOp;

import java.util.Collections;
import java.util.List;

/**
 * Backward pass for {@link BsrToDense}: gathers the upstream dense gradient back into BSR block values.
 *
 * <p>C++ op name: {@code bsr_to_dense_bp}
 *
 * <p><b>Inputs (3 arrays):</b>
 * <ol>
 *   <li>{@code bsrColIdx} – 1D [nnzb], INT — block column indices</li>
 *   <li>{@code bsrRowPtr} – 1D [mb+1], INT — block row pointers</li>
 *   <li>{@code gradDense} – 2D [rows, cols], float — upstream gradient w.r.t. dense output</li>
 * </ol>
 *
 * <p><b>Integer arguments (IArgs):</b> {@code rows}, {@code cols}, {@code blockDim}
 *
 * <p><b>Output (1 array):</b>
 * <ol>
 *   <li>{@code dBsrValues} – 1D [nnzb*blockDim*blockDim], same float dtype as {@code gradDense}</li>
 * </ol>
 *
 * <p>This op is a backward primitive: {@code doDiff} is not implemented.
 */
public class BsrToDenseBp extends DynamicCustomOp {

    /** No-arg constructor required for ImportClassMapping reflection. */
    public BsrToDenseBp() {}

    /**
     * Eager (INDArray) constructor.
     *
     * @param bsrColIdx 1D [nnzb] INT block column indices
     * @param bsrRowPtr 1D [mb+1] INT block row pointers
     * @param gradDense 2D [rows, cols] float upstream gradient w.r.t. dense output
     * @param rows      number of rows in the logical dense matrix
     * @param cols      number of columns in the logical dense matrix
     * @param blockDim  the square block size
     */
    public BsrToDenseBp(INDArray bsrColIdx, INDArray bsrRowPtr, INDArray gradDense,
                        long rows, long cols, long blockDim) {
        super(new INDArray[]{bsrColIdx, bsrRowPtr, gradDense}, null);
        addIArgument(rows, cols, blockDim);
    }

    /**
     * SameDiff (symbolic) constructor.
     *
     * @param sd        the SameDiff graph
     * @param bsrColIdx SDVariable [nnzb] INT block column indices
     * @param bsrRowPtr SDVariable [mb+1] INT block row pointers
     * @param gradDense SDVariable [rows, cols] float upstream gradient w.r.t. dense output
     * @param rows      number of rows in the logical dense matrix
     * @param cols      number of columns in the logical dense matrix
     * @param blockDim  the square block size
     */
    public BsrToDenseBp(SameDiff sd,
                        SDVariable bsrColIdx, SDVariable bsrRowPtr, SDVariable gradDense,
                        long rows, long cols, long blockDim) {
        super(sd, new SDVariable[]{bsrColIdx, bsrRowPtr, gradDense});
        addIArgument(rows, cols, blockDim);
    }

    @Override
    public String opName() {
        return "bsr_to_dense_bp";
    }

    @Override
    public int getNumOutputs() {
        return 1;
    }

    /**
     * Output data type: same float dtype as {@code gradDense} (input index 2).
     */
    @Override
    public List<DataType> calculateOutputDataTypes(List<DataType> dataTypes) {
        return Collections.singletonList(dataTypes.get(2));
    }
}

File: libnd4j/include/ops/declarable/generic/sparse/bsr_to_dense_bp.cpp
/* ******************************************************************************
 *
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 *  See the NOTICE file distributed with this work for additional
 *  information regarding copyright ownership.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 ******************************************************************************/

//
// bsr_to_dense_bp — backward pass for bsr_to_dense.
//
// Forward: BSR block at (br, bc = bsrColIdx[k]) is scattered into
//   dense[br*bd + lr, bc*bd + lc] = bsrValues[k*bd*bd + lr*bd + lc]
//
// Backward: gather from same region of the gradient:
//   dBsrValues[k*bd*bd + lr*bd + lc] = gradDense[br*bd + lr, bc*bd + lc]
//
// Inputs:
//   [0] bsrColIdx [nnzb]        INT   — BSR block column indices
//   [1] bsrRowPtr [mb+1]        INT   — BSR block row pointers (mb = rows/blockDim)
//   [2] gradDense [rows, cols]  float — upstream gradient w.r.t. dense output
// IArgs:
//   [0] rows
//   [1] cols
//   [2] blockDim
// Output:
//   [0] dBsrValues [nnzb*blockDim*blockDim] float — gradient w.r.t. BSR values
//

#include <system/op_boilerplate.h>
#if NOT_EXCLUDED(OP_bsr_to_dense_bp)

#include <ops/declarable/headers/sparse.h>
#include <ops/declarable/helpers/sparse_bsr_bp.h>

namespace sd {
namespace ops {

CUSTOM_OP_IMPL(bsr_to_dense_bp, 3, 1, false, 0, 3) {
  auto bsrColIdx = INPUT_VARIABLE(0);
  auto bsrRowPtr = INPUT_VARIABLE(1);
  auto gradDense = INPUT_VARIABLE(2);

  const LongType rows     = INT_ARG(0);
  const LongType cols     = INT_ARG(1);
  const LongType blockDim = INT_ARG(2);

  REQUIRE_TRUE(bsrColIdx->rankOf() == 1, 0,
               "bsr_to_dense_bp: bsrColIdx must be 1D, got rank %d", bsrColIdx->rankOf());
  REQUIRE_TRUE(bsrRowPtr->rankOf() == 1, 0,
               "bsr_to_dense_bp: bsrRowPtr must be 1D, got rank %d", bsrRowPtr->rankOf());
  REQUIRE_TRUE(gradDense->rankOf() == 2, 0,
               "bsr_to_dense_bp: gradDense must be 2D, got rank %d", gradDense->rankOf());
  REQUIRE_TRUE(gradDense->sizeAt(0) == rows && gradDense->sizeAt(1) == cols, 0,
               "bsr_to_dense_bp: gradDense shape must be [rows=%lld, cols=%lld], got [%lld, %lld]",
               (long long)rows, (long long)cols,
               (long long)gradDense->sizeAt(0), (long long)gradDense->sizeAt(1));

  const LongType nnzb = bsrColIdx->lengthOf();
  auto dBsrValues = OUTPUT_VARIABLE(0);

  if (nnzb > 0) {
    sd::ops::helpers::bsr_to_dense_bp(
        *bsrColIdx, *bsrRowPtr,
        *gradDense, *dBsrValues,
        rows, cols, blockDim);
  }

  return sd::Status::OK;
}

DECLARE_SHAPE_FN(bsr_to_dense_bp) {
  auto bsrColIdx = INPUT_VARIABLE(0);
  auto gradDense = INPUT_VARIABLE(2);

  const LongType nnzb     = bsrColIdx->lengthOf();
  const LongType blockDim = INT_ARG(2);
  const LongType len      = nnzb * blockDim * blockDim;
  auto outShape = ConstantShapeHelper::getInstance().createShapeInfo(
      gradDense->dataType(), 'c', {len});
  return SHAPELIST(outShape);
}

DECLARE_TYPES(bsr_to_dense_bp) {
  getOpDescriptor()
      ->setAllowedInputTypes(0, {ALL_INTS})     // bsrColIdx
      ->setAllowedInputTypes(1, {ALL_INTS})     // bsrRowPtr
      ->setAllowedInputTypes(2, {ALL_FLOATS})   // gradDense
      ->setAllowedOutputTypes(0, {ALL_FLOATS}); // dBsrValues
}

}  // namespace ops
}  // namespace sd

#endif  // NOT_EXCLUDED(OP_bsr_to_dense_bp)
